File: Compiler/Code/bench/multi_lang_benchmark/w3_matmul/matmul.cpp
#include <iostream>
#include <vector>
#include <chrono>

int main() {
    auto t0 = std::chrono::high_resolution_clock::now();
    const int n = 100;
    const int size = n * n;

    std::vector<int64_t> A(size, 0);
    std::vector<int64_t> B(size, 0);
    std::vector<int64_t> C(size, 0);

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            int idx = i * n + j;
            A[idx] = (i + j) % 10;
            B[idx] = (i * 2 + j) % 10;
        }
    }

    for (int i = 0; i < n; ++i) {
        for (int k = 0; k < n; ++k) {
            int64_t a_val = A[i * n + k];
            for (int j = 0; j < n; ++j) {
                C[i * n + j] += a_val * B[k * n + j];
            }
        }
    }

    int64_t checksum = 0;
    for (int i = 0; i < size; ++i) {
        checksum += C[i];
        if (checksum > 1000000000) {
            checksum %= 1000000000;
        }
    }

    auto t1 = std::chrono::high_resolution_clock::now();
    auto us = std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count();
    std::cout << "W3_MATMUL checksum=" << checksum << " time_us=" << us << "\n";
    return 0;
}

File: Compiler/Code/bench/multi_lang_benchmark/w3_matmul/matmul.rs
use std::time::Instant;

fn main() {
    let t0 = Instant::now();
    let n: usize = 100;
    let size = n * n;

    let mut a = vec![0i64; size];
    let mut b = vec![0i64; size];
    let mut c = vec![0i64; size];

    for i in 0..n {
        for j in 0..n {
            let idx = i * n + j;
            a[idx] = ((i + j) % 10) as i64;
            b[idx] = ((i * 2 + j) % 10) as i64;
        }
    }

    for i in 0..n {
        for k in 0..n {
            let a_val = a[i * n + k];
            for j in 0..n {
                c[i * n + j] += a_val * b[k * n + j];
            }
        }
    }

    let mut checksum: i64 = 0;
    for i in 0..size {
        checksum += c[i];
        if checksum > 1000000000 {
            checksum %= 1000000000;
        }
    }

    let us = t0.elapsed().as_micros();
    println!("W3_MATMUL checksum={} time_us={}", checksum, us);
}
